docs/design-proposals/multi-tenancy/multi-tenancy-overview.md

```diff
@@ -0,0 +1,149 @@
+
+A cluster is divided into control plane and data plane:
+
+ * On control plane side, Arktos manage isolate API resources from different tenants, and also isolate them from system API resources. Within each tenant scope, tenant admins can manage these API resources in an autonomous way.
+*  On data plane side, Arktos limits access within tenant scopes. For example, pod A from tenant A won't be able to access pod B from tenant B, even if tenant A users get the namespace, name or IP of pod B by guessing.
+
+Arktos achieves these based on a new concept "space". By default a tenant is allocated a space with the same tenant name, and only that tenant and system tenant has access to the space. All API object visibility and interactions are restricted to the space.
+
+By retrieving tenant credential from API request and automatically filling in the space information, we make spaces transparent to users. A tenant user accesses resources in its own space just like how he or she access Arktos without multi-tenancy. The APIs and tools are still compatible.
+
+
+## List of Component Changes
+
+### API Server
+
+* Implement tenant space "/tenants/{tenant}/resources"
+* Support short path for tenant space resources in endpoint handler.
+* Add a new API object "tenant" in system space.
+* Add a new API object "network" which can be in any space.
+* Admission controller to block certain tenant resources: node, daemonSet.
+* Admission controller: support space-level quota. 
+* Add tenantName in ObjectMeta type.
+* Modify bearer authenticator to extract tenant information.
+* Modify RBAC authorizer to check cross-tenant access, and skip clusterRoles not in the tenant scope.
+* API Discovery: return different resource G/V/K for the specific user
+* API Discovery: not return unavailable resource G/V/K like nodes, daemonSets
+* API Access: validate if the specified resources G/V/K is available for the given user
+
+### Scheduler
+* Use tenant.ALL to list/watch all resources.
+* Use tenant-specific path to update a specific resource.
+* (P2) Scheduling fairness: to avoid one tenant starve all other tenants.
+
+### Controllers
+* A new tenant controller to watch tenant objects, and perform initialization and cleanup work.
+* For all controllers: use tenant.ALL to list/watch all resources.
+* For all controllers: compare tenancy when do resource correlation.
+* For all controllers: use tenant-specific path to update a specific resource.
+* (P2) For all controllers: tenancy fairness.
+* A new flat network controller.
+* Modify the current network controller to VPC network controller.
+
+### Kubelet
+* Support Kata to provide container runtime isolation.
+* Set pod DNS based on the network the pod is attached to. 
+
+### Kube-proxy
+* Replaced by VPC LBs. (**TBD: require more details**)
+
+### Kubectl
+* Support CRUDW of the new API resource: api/v1/tenants.
+
```

Comment:
It can be supported by default tenants in client-go. So probably no code changes are required in Kubectl. Ideally users only need to set the default space in kubeconfig file, and they can access it using system tenant credential or user tenant credential.

 I don't suggest we add explicit -tenant option in kubectl for now. It probably will require too much work. We can put it P2 or P3 for now.

Reply:
Discussed in today's scrum and agreed to make client-go changes.